.github/actions/verify-token/src/utils.ts

```diff
@@ -22,3 +27,172 @@ export function getWorkflowPath(obj: githubObj): string {
+  core.debug(`tool-ref: ${toolRef}`);
+
+  // https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md
+  // We use the more recent Fulcio claims to extract the tool information.
+  // The commit sha for the tool is stored in
+  // a v3 extension with oid '1.3.6.1.4.1.57264.1.9'.
+  // https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726419--build-signer-uri
+  // `
+  //    1.3.6.1.4.1.57264.1.3:\n
+  //        8cbf4d422367d8499d5980a837cb9cc8e1e67001
+  // `
+  const textCertPath = "./client.txt";
+  const shaOid = "1.3.6.1.4.1.57264.1.10";
+  child_process
+    .execSync(
+      `openssl x509 -in ${clientCertPath} -noout -text -out ${textCertPath}`
+    )
+    .toString();
+  const resultSha = child_process
+    .execSync(`grep -A 1 '${shaOid}:' ${textCertPath}`)
+    .toString();
+  const indexSha = resultSha.indexOf(shaOid);
+  if (indexSha === -1) {
+    throw new Error(`error: cannot find oid '${shaOid}' in certificate`);
+  }
+
+  core.debug(`resultSha: ${resultSha}`);
+  const sha1ByteLen = 20;
+  const toolSha = resultSha
+    .slice(indexSha + `${shaOid}:`.length)
+    .replace("\n", "")
+    .trim()
+    .slice(-(sha1ByteLen * 2));
+  core.debug(`tool-sha: ${toolSha}`);
+
+  const toolPath = removeSuffix(
+    removePrefix(toolURI, `https://github.com/${toolRepository}/`),
+    `@${toolRef}`
+  );
+  core.debug(`tool-path: ${toolPath}`);
+  return [toolURI, toolRepository, toolRef, toolSha, toolPath];
+}
+
+function removeSuffix(s: string, suffix: string): string {
+  if (!s.endsWith(suffix)) {
+    throw new Error(`error: no suffix '${suffix}' in '${s}'`);
+  }
+  return s.slice(0, -suffix.length);
+}
+
+function removePrefix(s: string, prefix: string): string {
+  if (!s.startsWith(prefix)) {
+    throw new Error(`error: no prefix '${prefix}' in '${s}'`);
+  }
+  return s.slice(prefix.length);
+}
+
+function extractIdentifyFromSAN(URI: string): [string, string] {
+  // NOTE: the URI looks like:
+  // https://github.com/laurentsimon/slsa-delegated-tool/.github/workflows/tool1_slsa3.yml@refs/heads/main.
+  // We want to extract:
+  // - the repository: laurentsimon/slsa-delegated-tool
+  // - the ref: refs/heads/main
+  const parts = URI.split("@");
+  if (parts.length !== 2) {
+    throw new Error(`invalid URI (1): ${URI}`);
+  }
+  const ref = parts[1];
+  const url = parts[0];
+  const gitHubURL = "https://github.com/";
+  if (!url.startsWith(gitHubURL)) {
+    throw new Error(`not a GitHub URI: ${URI}`);
+  }
+  // NOTE: we omit the gitHubURL from the URL.
+  const parts2 = url.slice(gitHubURL.length).split("/");
+  if (parts2.length <= 2) {
+    throw new Error(`invalid URI (2): ${URI}`);
+  }
+  const repo = `${parts2[0]}/${parts2[1]}`;
+  return [repo, ref];
+}
+
+export async function fetchToolWorkflow(
+  ghToken: string,
+  repoName: string,
+  hash: string,
+  workflowPath: string
+): Promise<string> {
+  const url = `https://raw.githubusercontent.com/${repoName}/${hash}/${workflowPath}`;
```

Comment:
Can you not do this with the Octokit client that's part of the GitHub Actions toolkit?

I think it would be something like:

```typescript
import * as github from "@actions/github";

const octokit = github.getOctokit(ghToken);

let result = await octokit.repos.getContent({
  owner: ownerName,
  repo: repoName,
  path: workflowPath
})

// content will be base64 encoded
const body = Buffer.from(result.data.content, 'base64').toString()
```

Reply:
+1, the hash seems to be supported so I'll update. Thanks

Reply:
done, ptal